### PR TITLE
Migrate document classification code from castorini/castor

### PR DESCRIPTION
**Pull request by achyudh**
In preparation for the NAACL2019 camera ready and future work, I have made the following changes as a part of refactoring castorini/castor into castorini/hedwig: 
- Remove non-single input sequence models
- Move all models into models directory from root
- Declare common arguments in models/args.py, and move these arguments out of kim_cnn and reg_lstm 
- Removed baseline LSTM
- Renamed lstm_regularization into reg_lstm

### CODE REVIEW

File: models/char_cnn/args.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import os
+
```
Comment (achyudh) [inline models/char_cnn/args.py:2]: Fix imports according to PEP8. TBD in a later commit.

File: models/kim_cnn/__main__.py [hunk 2]
```diff
@@ -74,7 +70,7 @@ def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_si
         torch.cuda.set_device(args.gpu)
         torch.cuda.manual_seed(args.seed)
     if torch.cuda.is_available() and not args.cuda:
-        print('Warning: You have Cuda but not use it. You are using CPU for training.')
+        print('Warning: Using CPU for training')
```
Comment (daemon) [inline models/kim_cnn/__main__.py:73]: Should we use the `logging` module for everything?

File: models/reg_lstm/args.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    parser.add_argument('--mode', type=str, default='static', choices=['rand', 'static', 'non-static'])
+    parser.add_argument('--dataset', type=str, default='Reuters', choices=['SST-1', 'SST-2', 'Reuters', 'AAPD', 'IMDB', 'Yelp2014'])
+    parser.add_argument('--words-dim', type=int, default=300)
+    parser.add_argument('--embed-dim', type=int, default=300)
+    parser.add_argument('--dropout', type=float, default=0.5)
+    parser.add_argument('--epoch-decay', type=int, default=15)
+    parser.add_argument('--word-vectors-dir', default=os.path.join(os.pardir, 'Castor-data', 'embeddings', 'word2vec'))
+    parser.add_argument('--word-vectors-file', default='GoogleNews-vectors-negative300.txt')
+    parser.add_argument('--trained-model', type=str, default="")
+    parser.add_argument('--TAR', type=float, default=0.0, help="temporal activation regularization")
```
Comment (daemon) [inline models/reg_lstm/args.py]: `--tar` and `--ar` instead?

File: models/reg_lstm/__main__.py [hunk 4]
```diff
@@ -88,15 +86,20 @@ def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_si
...
+                                                                          unk_init=UnknownWordVecCache.unk)
 
     config = deepcopy(args)
     config.dataset = train_iter.dataset
     config.target_class = train_iter.dataset.NUM_CLASSES
     config.words_num = len(train_iter.dataset.TEXT_FIELD.vocab)
 
     print('Dataset {}    Mode {}'.format(args.dataset, args.mode))
-    print('VOCAB num',len(train_iter.dataset.TEXT_FIELD.vocab))
+    print('VOCAB num', len(train_iter.dataset.TEXT_FIELD.vocab))
```
Comment (daemon) [inline models/reg_lstm/__main__.py]: VOCAB?

File: models/reg_lstm/args.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    parser.add_argument('--resume-snapshot', type=str, default=None)
+    parser.add_argument('--bidirectional', action='store_true')
+    parser.add_argument('--bottleneck-layer', action='store_true')
+    parser.add_argument('--single-label', action='store_true')
+    parser.add_argument('--num-layers', type=int, default=2)
+    parser.add_argument('--hidden-dim', type=int, default=256)
+    parser.add_argument('--mode', type=str, default='static', choices=['rand', 'static', 'non-static'])
+    parser.add_argument('--dataset', type=str, default='Reuters', choices=['SST-1', 'SST-2', 'Reuters', 'AAPD', 'IMDB', 'Yelp2014'])
+    parser.add_argument('--words-dim', type=int, default=300)
```
Comment (daemon) [inline models/reg_lstm/args.py:17]: `--word-dim`?

File: models/han/args.py [hunk 6]
```diff
@@ -19,7 +19,6 @@ def get_args():
     parser.add_argument('--seed', type=int, default=3435)
     parser.add_argument('--dataset', type=str, default='SST-1', choices=['SST-1', 'SST-2', 'Reuters', 'AAPD', 'IMDB', 'Yelp2014'])
     parser.add_argument('--resume_snapshot', type=str, default=None)
-    parser.add_argument('--dev_every', type=int, default=30)
     parser.add_argument('--log_every', type=int, default=10)
```
Comment (daemon) [inline models/han/args.py:22]: `--log-every`, `--resume-snapshot`?

File: models/kim_cnn/args.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import os
+import models.args
+
+
+def get_args():
+    parser = models.args.get_args()
+
+    parser.add_argument('--dataset', type=str, default='SST-1', choices=['SST-1', 'SST-2', 'Reuters', 'AAPD', 'IMDB', 'Yelp2014'])
```
Comment (daemon) [inline models/kim_cnn/args.py:8]: Is this case sensitive?

File: models/reg_lstm/args.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import os
```
Comment (daemon) [inline models/reg_lstm/args.py:1]: \n

File: README.md [hunk 9]
```diff
@@ -4,18 +4,18 @@ This repo contains PyTorch deep learning models for document classification, imp
...
-+ [Conv-RNN](./conv_rnn/): Convolutional RNN [(Wang et al., KDD 2017)](https://dl.acm.org/citation.cfm?id=3098140)
-+ [HAN](./han/): Hierarchical Attention Networks [(Zichao, et al, NAACL 2016)](https://www.cs.cmu.edu/~hovy/papers/16HLT-hierarchical-attention-networks.pdf)
-+ [LSTM-Reg](./lstm_regularization/): Standard LSTM with Regularization [(Merity et al.)](https://arxiv.org/abs/1708.02182)
-+ [XML-CNN](./xml_cnn/): CNNs for Extreme Multi-label Text Classification [(Liu et al., SIGIR 2017)](http://nyc.lti.cs.cmu.edu/yiming/Publications/jliu-sigir17.pdf)
++ [Kim CNN](models/kim_cnn/): Baseline convolutional neural network for sentence classification [(Kim, EMNLP 2014)](http://www.aclweb.org/anthology/D14-1181)
++ [Conv-RNN](models/conv_rnn/): Convolutional RNN [(Wang et al., KDD 2017)](https://dl.acm.org/citation.cfm?id=3098140)
++ [HAN](models/han/): Hierarchical Attention Networks [(Zichao, et al, NAACL 2016)](https://www.cs.cmu.edu/~hovy/papers/16HLT-hierarchical-attention-networks.pdf)
++ [LSTM-Reg](models/reg_lstm/): Standard LSTM with Regularization [(Merity et al.)](https://arxiv.org/abs/1708.02182)
```
Comment (daemon) [inline README.md]: Regularization is neither a proper noun nor part of a title.

File: models/reg_lstm/args.py (hunk 5 above)
Comment (achyudh) [reply models/reg_lstm/args.py:17]: Hmm, it's consistently words-dim everywhere. I should change it in every model or I would be breaking some of them.

File: models/reg_lstm/__main__.py (hunk 4 above)
Comment (achyudh) [reply models/reg_lstm/__main__.py]: This part is sketchy in all of the main methods. I guess this is why it's not a great idea to have half of the methods in the code base copy-pasted huh~

File: models/han/args.py (hunk 6 above)
Comment (achyudh) [reply models/han/args.py:22]: I guess --log-every could be moved to the common args file but resume-snapshot is actually dependent on the task. For doc classification, it's loading a model whereas for relevance transfer it is loading the previous run for all topics. I am leaving the latter as it is.

File: models/kim_cnn/args.py (hunk 7 above)
Comment (achyudh) [reply models/kim_cnn/args.py:8]: Currently, I think so. Do you want everything to be in lowercase?

File: models/han/args.py (hunk 6 above)
Comment (achyudh) [reply models/han/args.py:22]: Oh wait, the reason it's still here for HAN is because I haven't started working on HAN yet. I'm fixng up `reg_lstm` and `kim_cnn` first.

File: README.md [hunk 10]
```diff
@@ -4,18 +4,18 @@ This repo contains PyTorch deep learning models for document classification, imp
 
 ## Models
 
-+ [Kim CNN](./kim_cnn/): Baseline convolutional neural network for sentence classification [(Kim, EMNLP 2014)](http://www.aclweb.org/anthology/D14-1181)
-+ [Conv-RNN](./conv_rnn/): Convolutional RNN [(Wang et al., KDD 2017)](https://dl.acm.org/citation.cfm?id=3098140)
-+ [HAN](./han/): Hierarchical Attention Networks [(Zichao, et al, NAACL 2016)](https://www.cs.cmu.edu/~hovy/papers/16HLT-hierarchical-attention-networks.pdf)
-+ [LSTM-Reg](./lstm_regularization/): Standard LSTM with Regularization [(Merity et al.)](https://arxiv.org/abs/1708.02182)
-+ [XML-CNN](./xml_cnn/): CNNs for Extreme Multi-label Text Classification [(Liu et al., SIGIR 2017)](http://nyc.lti.cs.cmu.edu/yiming/Publications/jliu-sigir17.pdf)
++ [Kim CNN](models/kim_cnn/): CNNs for sentence classification [(Kim, EMNLP 2014)](http://www.aclweb.org/anthology/D14-1181)
```
Comment (Ashutosh-Adhikari) [inline README.md:7]: Need to rewrite the README's

File: models/args.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import os
+from argparse import ArgumentParser
+
+
+def get_args():
+    parser = ArgumentParser(description="PyTorch deep learning models for document classification")
```
Comment (Ashutosh-Adhikari) [inline models/args.py:6]: Is this a single args file for all models combined? As in the model name needs to be passed, right?

File: README.md (hunk 10 above)
Comment (achyudh) [reply README.md:7]: Yes, I agree. Though this is something I would like to do after all the refactoring is done. Plus, we would have to write a runbook for NAACL too.

File: models/args.py (hunk 11 above)
Comment (Ashutosh-Adhikari) [reply models/args.py:6]: After this PR, explore inheritance for global and local args.

File: models/char_cnn/args.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    parser.add_argument('--num-conv-filters', type=int, default=256)
+    parser.add_argument('--num-affine-neurons', type=int, default=1024)
+    parser.add_argument('--output-channel', type=int, default=256)
+    parser.add_argument('--dropout', type=float, default=0.5)
+    parser.add_argument('--epoch-decay', type=int, default=15)
+    parser.add_argument('--data-dir', help='word vectors directory',
+                        default=os.path.join(os.pardir, 'Castor-data', 'datasets'))
+    parser.add_argument('--word-vectors-dir', help='word vectors directory',
+                        default=os.path.join(os.pardir, 'Castor-data', 'embeddings', 'word2vec'))
+    parser.add_argument('--word-vectors-file', help='word vectors filename', default='GoogleNews-vectors-negative300.txt')
```
Comment (Ashutosh-Adhikari) [inline models/char_cnn/args.py:29]: We should rename the word vectors file in Castor-data?

File: models/han/model.py [hunk 13]
```diff
@@ -1,9 +1,8 @@
 import torch
 import torch.nn as nn
-import torch.nn.functional as F
 
-from han.sent_level_rnn import SentLevelRNN
-from han.word_level_rnn import WordLevelRNN
+from models.han import SentLevelRNN
+from models.han import WordLevelRNN
```
Comment (Ashutosh-Adhikari) [inline models/han/model.py]: from models.han.sent_level_rnn import SentLevelRNN?

**Review by Ashutosh-Adhikari — CHANGES_REQUESTED** [review]

File: models/char_cnn/args.py (hunk 12 above)
Comment (achyudh) [reply models/char_cnn/args.py:29]: I was thinking about this too. At the very least it should be castor-data in lower case

File: models/args.py (hunk 11 above)
Comment (achyudh) [reply models/args.py:6]: If you are willing to make the changes, I would be glad to review them

File: models/han/model.py (hunk 13 above)
Comment (achyudh) [reply models/han/model.py]: I fixed this in the latest commit. Thanks for pointing it out.

**Review by Ashutosh-Adhikari — APPROVED** [review]
LGTM!
